### PR TITLE
[BuildSystem] Allow processes to release lanes

### PR DESCRIPTION
**Pull request by dmbryson**
The general approach here is that LaneBasedExecutionQueue::executeProcess()
exposes a pipe file descriptor to the launched process. If that process
sends the value of LLBUILD_TASK_ID over that pipe, the process will
release the lane on which it is executing and asynchronously wait for
completion.

This initial implementation spawns a new thread for each process that
triggers the release. Longer term, more sophisticated handling of async
IO and process wait may be preferred.

Since any process may opt in to use this functionality, the chain of
methods that lead into the executeProcess call have all been modified
to take completion/result reporting callbacks.

rdar://problem/44683736

### CODE REVIEW

**Comment by dmbryson** [comment]
@swift-ci please smoke test

**Comment by dmbryson** [comment]
@swift-ci please test

**Comment by dmbryson** [comment]
(not for merger yet, I plan to do some local ubuntu 14.04 tests before doing so)

File: docs/buildsystem.rst [hunk 1]
```diff
@@ -611,6 +611,40 @@ the output file -- even if the output file was just regenerated. This is under
...
+
+   * - Variable
+     - Description
+
+   * - LLBUILD_LANE_ID
+     - An identifier string representing the execution lane the task is running
+       under.
+
+   * - LLBUILD_TASK_ID
+     - A (reasonably) identifier string for this individual task.
```
Comment (ddunbar) [inline docs/buildsystem.rst]: Reasonably what?

Comment (dmbryson) [reply docs/buildsystem.rst]: Heh "unique"

**Comment by dmbryson** [comment]
@swift-ci please smoke test

**Comment by dmbryson** [comment]
@swift-ci please smoke test

**Comment by ddunbar** [comment]
🎉🎉🎉

🤞🤞🤞

**Pull request merged**